cloud-controller-manager/do/firewall_controller.go

```diff
@@ -117,35 +125,42 @@ func NewFirewallController(
 		workerFirewallName: workerFirewallName,
 		fwManager:          fwManager,
 		metrics:            metrics,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "firewall"),
 	}
 
 	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: func(cur interface{}) {
-				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
-				defer cancel()
-
-				err := fc.observeReconcileDuration(ctx, originEventAdd)
-				if err != nil {
-					klog.Errorf("failed to reconcile worker firewall after adding a resource object: %s", err)
+				key, err := cache.MetaNamespaceKeyFunc(cur)
+				if err == nil {
+					fc.queue.Add(key)
+
+					reconcileErr := fc.observeReconcileDuration(ctx, originEventAdd)
+					if reconcileErr != nil {
+						klog.Errorf("failed to reconcile worker firewall after adding a resource object: %s", err)
+					}
 				}
 			},
 			UpdateFunc: func(old, cur interface{}) {
-				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
-				defer cancel()
-
-				err := fc.observeReconcileDuration(ctx, originEventUpdate)
-				if err != nil {
-					klog.Errorf("failed to reconcile worker firewall after updating a resource object: %s", err)
+				key, err := cache.MetaNamespaceKeyFunc(cur)
+				if err == nil {
+					fc.queue.Add(key)
+
+					reconcileErr := fc.observeReconcileDuration(ctx, originEventUpdate)
+					if reconcileErr != nil {
+						klog.Errorf("failed to reconcile worker firewall after updating a resource object: %s", err)
+					}
 				}
 			},
-			DeleteFunc: func(obj interface{}) {
-				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
-				defer cancel()
-
-				err := fc.observeReconcileDuration(ctx, originEventDelete)
-				if err != nil {
-					klog.Errorf("failed to reconcile worker firewall after deleting a resource object: %s", err)
+			DeleteFunc: func(cur interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(cur)
```

Comment:
I think we should use `DeletionHandlingMetaNamespaceKeyFunc` as in the [official example](https://github.com/kubernetes/client-go/blob/ba26c4506d39dece49d00e969d03794e8089b8dc/examples/workqueue/main.go#L188-L190).

Reply:
Nice catch!